src/solvers/solver_utils.hpp

```diff
@@ -135,6 +145,155 @@ struct Stencil {
+
+  int nblocks = md->NumBlocks();
+  std::vector<bool> include_block(nblocks, true);
+  if (only_interior) {
+    // The neighbors array will only be set for a block if its a leaf block
+    for (int b = 0; b < nblocks; ++b)
+      include_block[b] = md->GetBlockData(b)->GetBlockPointer()->neighbors.size() == 0;
+  }
+
+  auto desc = parthenon::MakePackDescriptor<a_t, b_t, out>(md.get());
+  auto pack = desc.GetPack(md.get(), include_block, only_fine_on_composite);
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "AddFieldsAndStore", DevExecSpace(), 0, pack.GetNBlocks() - 1,
+      kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+        const int nvars = pack.GetUpperBound(b, a_t()) - pack.GetLowerBound(b, a_t()) + 1;
+        for (int c = 0; c < nvars; ++c) {
+          pack(b, te, out(c), k, j, i) =
+              wa * pack(b, te, a_t(c), k, j, i) + wb * pack(b, te, b_t(c), k, j, i);
+        }
+      });
+  return TaskStatus::complete;
+}
+
+template <class a_t, class b_t, class out, bool only_fine_on_composite = true>
+TaskStatus AddFieldsAndStore(const std::shared_ptr<MeshData<Real>> &md, Real wa = 1.0,
+                             Real wb = 1.0) {
+  return AddFieldsAndStoreInteriorSelect<a_t, b_t, out, only_fine_on_composite>(
+      md, wa, wb, false);
+}
+
+template <class var, bool only_fine_on_composite = true>
+TaskStatus SetToZero(const std::shared_ptr<MeshData<Real>> &md) {
+  int nblocks = md->NumBlocks();
+  using TE = parthenon::TopologicalElement;
+  TE te = TE::CC;
+  std::vector<bool> include_block(nblocks, true);
+  auto desc = parthenon::MakePackDescriptor<var>(md.get());
+  auto pack = desc.GetPack(md.get(), include_block, only_fine_on_composite);
+  const size_t scratch_size_in_bytes = 0;
+  const int scratch_level = 1;
+  const int ng = parthenon::Globals::nghost;
+  parthenon::par_for_outer(
+      DEFAULT_OUTER_LOOP_PATTERN, "SetFieldsToZero", DevExecSpace(),
+      scratch_size_in_bytes, scratch_level, 0, pack.GetNBlocks() - 1,
+      KOKKOS_LAMBDA(parthenon::team_mbr_t member, const int b) {
+        auto cb = GetIndexShape(pack(b, te, 0), ng);
+        const auto &coords = pack.GetCoordinates(b);
+        IndexRange ib = cb.GetBoundsI(IndexDomain::interior, te);
+        IndexRange jb = cb.GetBoundsJ(IndexDomain::interior, te);
+        IndexRange kb = cb.GetBoundsK(IndexDomain::interior, te);
+        parthenon::par_for_inner(parthenon::inner_loop_pattern_simdfor_tag, member, kb.s,
+                                 kb.e, jb.s, jb.e, ib.s, ib.e, [&](int k, int j, int i) {
+                                   const int nvars = pack.GetUpperBound(b, var()) -
+                                                     pack.GetLowerBound(b, var()) + 1;
+                                   for (int c = 0; c < nvars; ++c)
+                                     pack(b, te, var(c), k, j, i) = 0.0;
+                                 });
```

Comment:
Yeah, I agree that it would be better to wrap the `par_for_inner` as 
```c++
const int nvars = pack.GetUpperBound(b, var()) -
                             pack.GetLowerBound(b, var()) + 1;
for (int c = 0; c < nvars; ++c) {
  parthenon::par_for_inner(...) {
      pack(b, te, var(c), k, j, i) = 0.0;
    });
}
```
and something similar below. It was written initially without accounting for the fact that variables could be more than size one, and I just quickly hacked this in there. Currently, this doesn't have any noticeable impact on performance.

Reply:
I fixed this in one place where there was already hierarchical parallelism being used. Other places in utils I just added a todo saying the change could be made if these kernels were impacting performance.